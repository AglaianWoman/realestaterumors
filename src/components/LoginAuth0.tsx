
import * as React from 'react'
import gql from 'graphql-tag'
import { graphql, compose } from 'react-apollo'
import { connect } from 'react-redux'

import { Redirect } from 'react-router-dom'
import Auth0Lock from 'auth0-lock'

import { userGQL } from './interfaceDefinitions'
import 'styles/LoginAuth0.scss'

import Title from './Title'
import { SpinnerRectangle, SpinnerDots } from './Spinners'

import * as Button from 'antd/lib/button'
import 'antd/lib/button/style/css'


interface LoginAuth0Props {
  clientId: string
  domain: string
  createUser(): void
  redirectOnAuth: string
  updateUserProfileRedux(userProfile: userGQL): void
  data?: {
    loading?: boolean
    error?: boolean | string
    user?: userGQL
  }
}

export class LoginAuth0 extends React.Component<LoginAuth0Props, any> {

  constructor(props) {
    super(props)
    this.lock = new Auth0Lock(this.props.clientId, this.props.domain)
    this.state = {
      redirect: false
    }
  }

  componentDidMount() {

    this.lock.on('authenticated', ({ accessToken, idToken }) => {
      window.localStorage.setItem('auth0IdToken', idToken)

      this.lock.getProfile(idToken, (err, profile) => {
        window.localStorage.setItem('profile', JSON.stringify(profile))
      })
      console.info("Authenticated!: ", window.localStorage.getItem('auth0IdToken'))

      var promise = new Promise((resolve, reject) => {
        resolve(this.props.data.refetch())
      })
      .then(res => {
        if (this.props.data.user) {
          this.props.updateUserProfileRedux(this.props.data.user)
        } else {
          console.info("New user! Making a new GraphCool User account.")
          this.createUser()
          this.props.data.refetch()
        }
      })
      .then(res => this.setState({ redirect: true }))
      .catch(err => console.warn(err))

    })

    this.lock.on('authorization_error', authResult => {
      console.log(authResult)
    })

  }

  createUser = () => {
    const variables = {
      idToken: window.localStorage.getItem('auth0IdToken'),
      emailAddress: JSON.parse(window.localStorage.getItem('profile')).email,
      emailVerified: JSON.parse(window.localStorage.getItem('profile')).email_verified,
      name: JSON.parse(window.localStorage.getItem('profile')).name,
    }
    this.props.createUser({ variables })
    .catch(err => console.log(err))
  }

  showLogin = () => {
    this.lock.show()
  }

  logOut = () => {
    window.localStorage.removeItem('auth0IdToken')
    window.localStorage.removeItem('profile')
    this.props.data.refetch()
  }

  render() {
    if (this.props.data.loading) {
      return (
        <div className='login-auth0'>
          <Button id='antd-login'>
            <div className='login-auth0-loader'>
              <SpinnerRectangle height='12px' width='4px'/>
              Loading
            </div>
          </Button>
        </div>
      )
    }
    if (this.props.data.user) {
      return (
        <div className='login-auth0'>
          <Button id='antd-login' onClick={this.logOut}>
            Log Out
          </Button>
          {(
            this.state.redirect &&
            <Redirect to={this.props.redirectOnAuth}/>
          )}
        </div>
      )
    } else {
      return (
      <div className='login-auth0'>
        <Button id='antd-login' type="primary" onClick={this.showLogin}>
          Login
          <Redirect to="/"/>
        </Button>
      </div>
      )
    }
  }
}




const CreateUserQuery = gql`
mutation ($idToken: String!, $name: String!, $emailAddress: String!) {
  createUser(authProvider: {auth0: {idToken: $idToken}}, name: $name, emailAddress: $emailAddress) {
    id
  }
}
`
// To check if a request is actually authenticated: user query
const UserQuery = gql`
query {
  user {
    id
    emailAddress
    predictions {
      id
      prediction
      house {
        id
        address
        unitNum
        streetNum
        streetName
        streetType
        lotPlan
      }
    }
  }
}
`

const mapDispatchToProps = ( dispatch ) => {
  return {
    updateUserProfileRedux: (userProfile) => dispatch({ type: "USER_GQL", payload: userProfile }),
  }
}

export default compose(
  connect(null, mapDispatchToProps), // connect dispatch to redux
  graphql(UserQuery, { options: { fetchPolicy: 'network-only' } }),
  // graphql(UserQuery),
  graphql(CreateUserQuery, { name: 'createUser' }),
)( LoginAuth0 )

