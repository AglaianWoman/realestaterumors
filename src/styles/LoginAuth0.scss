

.login-auth0 {
  position: fixed;
  top: 10px;
  right: 10px;
  z-index: 2;
}

.login-auth0-loader {
  color: #eee;
  display: flex;
  justify-content: center;
  width: 40px;
  transition: width 0.6s ease-in-out;
}
.login-auth0-loader.login-loading {
  width: 70px;
  transition: width 0.6s ease-in-out;
}
.login-loading::after {
  content: 'Loading';
}
.login-logged-in::after {
  content: 'Log Out';
}

button#antd-login {
  height: 32px;
  border-radius: 0px;
  background-color: #96616B;
  border-color: lighten(#96616B, 10%);
  color: #eee;
}

button#antd-login:hover {
  background-color: #d4afdf;
  border-color: #d4afdf;
}


// CSS transitions
// 1st time appearing
$min-width: 0px;
$max-width: 65px;
.login-anim-fade-appear {
  opacity: 0.01;
  width: $min-width;
}
.login-anim-fade-appear.login-anim-fade-appear-active {
  opacity: 1;
  transition: opacity 200ms ease-in;
  width: $max-width;
  transition: width 200ms ease-in;
}
// entering
.login-anim-fade-enter {
  opacity: 0.01;
  width: $min-width;
}
.login-anim-fade-enter.login-anim-fade-enter-active {
  opacity: 1;
  width: $max-width;
  transition: opacity 200ms ease-in;
  transition: width 200ms ease-in;
}
// leaving
.login-anim-fade-leave {
  opacity: 1;
  width: $max-width;
}
.login-anim-fade-leave.login-anim-fade-leave-active {
  opacity: 0.01;
  transition: opacity 200ms ease-in;
  width: $min-width;
  transition: width 200ms ease-in;
}


