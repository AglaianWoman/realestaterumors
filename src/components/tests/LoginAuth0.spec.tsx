
import * as React from 'react'
import * as renderer from 'react-test-renderer'
import { mount, shallow } from 'enzyme'

import { LoginAuth0 } from '../LoginAuth0'
import * as Button from 'antd/lib/button'


const mockData = {
  loading: false
  user: {'id': 'test_test_test'}
}


const MockProps = {
  clientId: 'test',
  domain: 'test',
  data: {
    loading: false,
    user: { 'id': 'mock ID #1234' }
  }
  updateUserProfileRedux: jest.fn()
}

test('<LoginAuth0 /> component should contain button', () => {
  beforeEach(() => {
    LoginAuth0.contextTypes = {
      router: jest.fn()
    }
    const el = shallow(
      <LoginAuth0 {...MockProps} />
    )
    expect(el.find(Button).length).toEqual(1)
  })
})


test('<LoginAuth0 /> component should render snapshot', () => {
  beforeEach(() => {
    LoginAuth0.contextTypes = {
      router: jest.fn()
    }
    const elR = renderer.create(
      <LoginAuth0 {...MockProps} />
    )
    expect(elR.toJSON()).toMatchSnapshot()
  })
})


