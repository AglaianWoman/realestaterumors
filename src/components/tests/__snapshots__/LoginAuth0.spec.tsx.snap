// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginAuth0 /> component should render snapshot 1`] = `
<div
  className="login-auth0"
>
  <button
    id="red-login-button"
    onClick={[Function]}
  >
    <div
      className="login-auth0-loader login-logged-in"
    />
  </button>
</div>
`;
